### ISSUE TITLE
Integrate Neo4j as a Graph Index, Vector Index, and as tools in the ecosystem

### ISSUE THREAD

**Issue by quillan86** [hwchase17/langchain#4625]
### Feature request

There is a need for graph databases to be integrated in langchain. NetworkX isn't suitable for scalable graph databases that would be desired to be queried, particularly with tens of thousands or more nodes and edges. This is necessary for graph databases to compete with vector databases on the level for information extraction within langchain.

There is already a [medium article](https://towardsdatascience.com/integrating-neo4j-into-the-langchain-ecosystem-df0e988344d2) and [GitHub repo](https://github.com/tomasonjo/langchain2neo4j) talking about one way in which this is implemented, but it would be ideal if something like this was integrated into langchain itself. This implementation also has Neo4j as embeddings as an option, which should be implemented as well.

### Motivation

The [Graph Index Creator](https://python.langchain.com/en/latest/modules/chains/index_examples/graph_qa.html?highlight=GraphIndexCreator) and other small forms of graphs within LangChain use NetworkX which isn't scalable for production for full blown knowledge graphs on the size of the vector databases. I know that I have a particular need to use a graph database in production along with langchain due to a work level project.

### Your contribution

Yes, I am willing to contribute. I haven't contributed to LangChain directly before but I am familiar with the source code. Would love to collaborate on what kind of framework/interface we would need to expand graph indexes with a similar scope as vector database indexes.

**Comment by tomasonjo** [comment]
I would also be willing to contribute, I would just need a bit of help to know where to put the code?
The closest sections seems vector store, but Neo4j is not a vector store, so should it be a retrieval or a tool, or do we just pretend neo4j is a vector store?

**Comment by quillan86** [comment]
There's already these folders of relevance:

- `langchain.graphs` (storing a barebones networkx graph)
- `langchain.indexes` (storing `graphs.py` for the GraphIndexCreator much liek the VectorStore Creator)
- `langchain.chains.graph_qa` (storing a chain for graph QA)

So I think it's a matter of reformulating `langchain.graphs` to have a base.py et al similar to `langchain.vectorstores`. That's why I said interface - it would be the creation of a new general object like `Vectorstore`.

We can possibly store the vector embedding portion of Neo4j within the vectorstore one, though, but I'd need to look at the code based on the medium article.

**Comment by quillan86** [comment]
Yeah that wouldn't be a priority atm (other than that was a feature) - cypher search would be the priority.

**Comment by tomasonjo** [comment]
I've started the PR, you can take a look

**Comment by tomasonjo** [comment]
This was added, so you could probably close this issue:

https://github.com/hwchase17/langchain/blob/master/docs/modules/chains/examples/graph_cypher_qa.ipynb

**Comment by v-almonacid** [comment]
> I've started the PR, you can take a look

@tomasonjo Quick question: Does `GraphCypherQAChain` works well for you? If yes, what version?

I tried the example in the docs with the current latest version (`0.0.197`) but it throws.

**Comment by tomasonjo** [comment]
Whats the error you are getting?

**Comment by v-almonacid** [comment]
> Whats the error you are getting?

I get the issue now. The LLM simply doesn't respond with a plain Cypher statement, so naturally `Neo4jGraph.query()` fails.
Maybe it's because I'm using an Azure LLM instance and it doesn't behave the same (?)

**Comment by tomasonjo** [comment]
I dont have access to azure llms, so I can't test it. You can ask the llm to wrap the statement in three backticks as the code can extract the statement then
